pkg/ccl/changefeedccl/cdcevent/event_test.go

```diff
@@ -134,6 +139,147 @@ CREATE TABLE foo (
 	}
 }
 
+// TestEventDescriptorWithSchemaChanges validates that complex schema changes
+// will produce valid event descriptors.q
+func TestEventDescriptorWithSchemaChanges(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	var hookEnabled atomic.Bool
+	var validationFn func(stage string) error
+	srv, db, _ := serverutils.StartServer(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			SQLDeclarativeSchemaChanger: &scexec.TestingKnobs{
+				BeforeStage: func(p scplan.Plan, stageIdx int) error {
+					if !hookEnabled.Load() {
+						return nil
+					}
+					return validationFn(fmt.Sprintf("%s/stage=%d", p.Params.ExecutionPhase, stageIdx))
+				},
+			},
+		},
+	})
+	defer srv.Stopper().Stop(context.Background())
+	s := srv.ApplicationLayer()
+
+	sqlDB := sqlutils.MakeSQLRunner(db)
+	sqlDB.Exec(t, `CREATE TYPE status AS ENUM ('open', 'closed', 'inactive')`)
+	sqlDB.Exec(t, `
+CREATE TABLE foo (
+  a INT,
+  b STRING,
+  c STRING NOT NULL,
+  d STRING AS (concat(b, c)) VIRTUAL,
+  e status NOT NULL,
+  PRIMARY KEY (b, a),
+  FAMILY main (a, b, e),
+  FAMILY only_c (c)
+)`)
+
+	for _, tc := range []struct {
+		schemaChange    string
+		expectedKeyCols [][]string
+		expectedColumns [][]string
+		expectedUDTCols [][]string
+	}{
+		{
+			// Replace the primary index with one that contains a hash.
+			schemaChange:    "ALTER TABLE foo ALTER PRIMARY KEY USING COLUMNS(b, a) USING HASH",
+			expectedKeyCols: [][]string{{"b", "a"}, {"crdb_internal_a_b_shard_16", "b", "a"}},
+			expectedColumns: [][]string{{"a", "b", "e"}, {"a", "b", "e"}},
+			expectedUDTCols: [][]string{{"e"}, {"e"}},
+		},
+		{
+			// Replace the primary index with a new one with a hash. This will lead
+			// to 3 indexes observed:
+			// 1) The original from the previous test
+			// 2) An index without the hash and just c + e
+			// 3) An index with c + e and a new hash.
+			schemaChange:    "ALTER TABLE foo ALTER PRIMARY KEY USING COLUMNS(c, e) USING HASH",
+			expectedKeyCols: [][]string{{"crdb_internal_a_b_shard_16", "b", "a"}, {"c", "e"}, {"crdb_internal_c_e_shard_16", "c", "e"}},
+			expectedColumns: [][]string{{"a", "b", "e"}, {"a", "b", "e"}, {"a", "b", "e"}},
+			expectedUDTCols: [][]string{{"e"}, {"e"}, {"e"}},
```

Comment:
I'm having a hard time understanding why these columns are the way they are - why is `{"a", "b", "e"}` repeated 3 times? what's with the internal_shard thing? can you add some comments to explain this a little bit please?